iceoryx_hoofs/include/iceoryx_hoofs/internal/containers/uninitialized_array.hpp

```diff
@@ -0,0 +1,117 @@
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_CONTAINERS_UNINITIALIZED_ARRAY_HPP
+#define IOX_HOOFS_CONTAINERS_UNINITIALIZED_ARRAY_HPP
+
+#include "iceoryx_hoofs/iceoryx_hoofs_types.hpp"
+
+#include <cstdint>
+
+namespace iox
+{
+namespace containers
+{
+/// @brief struct used as policy parameter in UninitializedArray to wrap an array with all elements zeroed
+/// @tparam ElementType is the array type
+/// @tparam Capacity is the array size
+template <typename ElementType, uint64_t Capacity>
+struct ZeroedBuffer
+{
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    // NOLINTBEGIN(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+    using element_t = cxx::byte_t[sizeof(ElementType)];
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    alignas(ElementType) element_t value[Capacity]{};
+    // NOLINTEND(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+};
+
+/// @brief struct used as policy parameter in UninitializedArray to wrap an uninitialized array
+/// @tparam ElementType is the array type
+/// @tparam Capacity is the array size
+template <typename ElementType, uint64_t Capacity>
+struct NonZeroedBuffer
+{
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    // NOLINTBEGIN(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+    using element_t = cxx::byte_t[sizeof(ElementType)];
+    // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required by low level UninitializedArray building block and encapsulated in abstraction
+    alignas(ElementType) element_t value[Capacity];
+    // NOLINTEND(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+};
+
+/// @brief Wrapper class for a C-style array of type ElementType and size Capacity. Per default it is uninitialized but
+/// all elements can be zeroed via template parameter ZeroedBuffer.
+/// @tparam ElementType is the array type
+/// @tparam Capacity is the array size
+/// @tparam Buffer is the policy parameter to choose between an uninitialized, not zeroed array (=NonZeroedBuffer,
+/// default) and an uninitialized array with all elements zeroed (=ZeroedBuffer)
+/// @note Out of bounds access leads to undefined behavior
+template <typename ElementType, uint64_t Capacity, template <typename, uint64_t> class Buffer = NonZeroedBuffer>
+class UninitializedArray
+{
+    static_assert(Capacity > 0U, "The size of the UninitializedArray must be greater than 0!");
+
+  public:
+    using value_type = ElementType;
+    using iterator = ElementType*;
+    using const_iterator = const ElementType*;
+
+    constexpr UninitializedArray() noexcept = default;
+    UninitializedArray(const UninitializedArray&) = delete;
```

Comment:
Not being able to copy it might be problematic for use in classes that should be copied.  For trivially copyable types T this could even use memcpy. We can defer this to another issue and solve it when needed.

Resolve or create issue and resolve.

Reply:
IMO the issue should be created when we encounter this problem.